### PR TITLE
Implement computeDryBaseline for RenderChip

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Also, use `getDryLayout` for the first pass of the `label` layout.  

The computeDryBaseline stuff doesn't have a test because there're asserts making sure it returns the same result as `computeDistanceToActualBaseline`. I have tried adding `+1` to the return expression and `flutter test` fails.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -1830,25 +1819,36 @@ class _RenderChip extends RenderBox with SlottedContainerRenderObjectMixin<_Chip
...
+        + theme.padding.top + theme.labelPadding.top;
+    return labelBaseline.offset;
+  }
+
   _ChipSizes _computeSizes(BoxConstraints constraints, ChildLayouter layoutChild) {
     final BoxConstraints contentConstraints = constraints.loosen();
     // Find out the height of the label within the constraints.
-    final Offset densityAdjustment = Offset(0.0, theme.visualDensity.baseSizeAdjustment.dy / 2.0);
-    final Size rawLabelSize = layoutChild(label, contentConstraints);
+    final Size rawLabelSize = label.getDryLayout(contentConstraints);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/chip.dart:1834]: The label will be laid out a second time in this method, with a set of different constraints. Using `getDryLayout` prevent the child from actually changing size and creating other side effects. The downside of using `getDryLayout` is that you won't be able to use `LayoutBuilder`

File: packages/flutter/lib/src/material/chip.dart [hunk 2]
```diff
@@ -1830,25 +1819,36 @@ class _RenderChip extends RenderBox with SlottedContainerRenderObjectMixin<_Chip
...
     final double contentSize = math.max(
       _kChipHeight - theme.padding.vertical + theme.labelPadding.vertical,
       rawLabelSize.height + theme.labelPadding.vertical,
     );
+    assert(contentSize >= rawLabelSize.height);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/chip.dart:1839]: we have been assuming that `contentSize >= rawLabelHeight`: 
```dart
    final BoxConstraints labelConstraints = BoxConstraints(
      minHeight: rawSize.height,
      maxHeight: size.height,
      maxWidth: maxLabelWidth,
    );
```
so I'm just making it more explicit

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by LongCatIsLooong** [comment]
> The _RenderDecoration class does not support dry layout.

Looks like this will have to wait until `_RenderDecoration.computeDryBaseline` is implemented.

**Pull request merged**